### PR TITLE
Speed up retraining times  approximately 2x

### PR DESCRIPTION
**Pull request by sergejsrk**
Speed up is achieved by:
1. saving (and especially loading) bottleneck arrays in binary files instead of .txt files
2. caching already calculated bottleneck arrays in RAM, instead of reading them every time from disk

P.S.
2x speed up is achieved on SSD disks, on HDD disks speed up is even greater

### CODE REVIEW

**Comment by sergejsrk** [comment]
I signed it!

**Comment by martinwicke** [comment]
Could you reformat this to match pep8, except for the indent, which should be 2 spaces?

File: tensorflow/examples/image_retraining/retrain.py [hunk 1]
```diff
@@ -371,6 +372,44 @@ def ensure_dir_exists(dir_name):
...
+
+  Args:
+    list_of_floats: List of floats we want to write to a file.
+    file_path: Path to a file where list of floats will be stored.
+
+  """
+
+  s = struct.pack('d' * BOTTLENECK_TENSOR_SIZE, *list_of_floats)
+
+  f = open(file_path, 'wb')
```
Comment (Mistobaan) [inline tensorflow/examples/image_retraining/retrain.py]: use `with open(fpath) as f:`

File: tensorflow/examples/image_retraining/retrain.py [hunk 2]
```diff
@@ -371,6 +372,44 @@ def ensure_dir_exists(dir_name):
...
+  """Reads list of floats from a given file.
+
+  Args:    
+    file_path: Path to a file where list of floats was stored.
+  Returns:
+    Array of bottleneck values (list of floats).
+
+  """
+
+  f = open(file_path, 'rb')
```
Comment (Mistobaan) [inline tensorflow/examples/image_retraining/retrain.py]: use `with open(fpath) as f:`

File: tensorflow/examples/image_retraining/retrain.py [hunk 3]
```diff
@@ -371,6 +372,44 @@ def ensure_dir_exists(dir_name):
...
+  s = []
+  s = struct.unpack('d' * BOTTLENECK_TENSOR_SIZE, f.read())
+  f.close()
+  # print (list(s))
```
Comment (Mistobaan) [inline tensorflow/examples/image_retraining/retrain.py]: remove comment

**Comment by sergejsrk** [comment]
Reformated according to PEP8

**Comment by sergejsrk** [comment]
One more thing considering creation of bottlenecks. 

If a new version of retrain.py is run in a bottleneck directory, that contains bottlenecks files created with a previous version of retrain.py then retrain.py will crash (it will try to read list of floats from .txt file and fail). 

To prevent this problem we can either:

- delete all bottlenecks created with a previous version of retrain.py (it is remove .txt files from bottleneck dir)
- or modify get_bottleneck_path function, so it appends not ".txt", but ".bin" postfix.

Which one is better option?

**Comment by danmane** [comment]
@petewarden can you review this please?

**Comment by petewarden** [comment]
Jenkins, test this please.

**Comment by petewarden** [comment]
I've reviewed the changes, and they look good to me! Thanks for contributing this @sergejsrk

**Pull request merged**